backend/src/main/java/org/eclipse/tractusx/puris/backend/common/edc/logic/service/EdcAdapterService.java

```diff
@@ -830,78 +833,120 @@ private boolean initiateDtrTransferForLookupApi(Partner partner, String specific
                         var resultObject = aasJson.get("result").get(0);
                         var submodelDescriptors = resultObject.get("submodelDescriptors");
-
-                        var submodelDescriptor = submodelDescriptors.get(0);
-                        var semanticId = submodelDescriptor.get("semanticId");
-                        var keys = semanticId.get("keys");
-                        for (var key : keys) {
-                            var keyType = key.get("type").asText();
-                            var keyValue = key.get("value").asText();
-                            if ("GlobalReference".equals(keyType) && "urn:samm:io.catenax.item_stock:2.0.0#ItemStock".equals(keyValue)) {
-                                var endpoints = submodelDescriptor.get("endpoints");
-                                var endpoint = endpoints.get(0);
-                                var interfaceObject = endpoint.get("interface").asText();
-                                if ("SUBMODEL-3.0".equals(interfaceObject)) {
-                                    var protocolInformationObject = endpoint.get("protocolInformation");
-                                    String href = protocolInformationObject.get("href").asText();
-                                    var findMaterialNumber = HttpUrl.parse(href).encodedPathSegments().stream().filter(segment -> urnPattern.matcher(segment).matches()).findFirst();
-                                    if (findMaterialNumber.isEmpty()) {
-                                        log.error("Failed to find materialNumber segment in HREF URL");
-                                        return false;
+                        for (var submodelDescriptor : submodelDescriptors) {
+                            var semanticId = submodelDescriptor.get("semanticId");
+                            var keys = semanticId.get("keys");
+                            for (var key : keys) {
+                                var keyType = key.get("type").asText();
+                                var keyValue = key.get("value").asText();
+                                if ("GlobalReference".equals(keyType) && "urn:samm:io.catenax.item_stock:2.0.0#ItemStock".equals(keyValue)) {
+                                    var endpoints = submodelDescriptor.get("endpoints");
+                                    var endpoint = endpoints.get(0);
+                                    var interfaceObject = endpoint.get("interface").asText();
+                                    if ("SUBMODEL-3.0".equals(interfaceObject)) {
+                                        var protocolInformationObject = endpoint.get("protocolInformation");
+                                        String href = protocolInformationObject.get("href").asText();
+                                        var findMaterialNumber = HttpUrl.parse(href).encodedPathSegments().stream().filter(segment -> urnPattern.matcher(segment).matches()).findFirst();
```

Comment:
Hmm, that's an interesting idea. But that would cause some difficulties. 

For example, if we would do it this way in the case of the ItemStock submodel. 

If the EdcContractMappingService would be handed over the ShellDescriptor. Then this Service would not know, under which mapping key it would have to store the href information. That information would have to be additionally provided from the EdcAdapterService method in an accompanying parameter. 

However in the case of the PartTypeInformation Submodel, the key is always the partnerMaterialNumber. 

So in essence, we always have to consider the individual traits of our submodels. IMHO, there is unfortunately no easy way to create an abstraction that would work for any submodel ...

Reply:
I would start with extract the whole part as something like `extractContractMappingsFromShellDescriptor` with parameters ShellDescriptor and mpr. Currently, we always update all ContractMappings and we can do it the same way in the new method. 

Benefit would be that the ContractMapping logic is further centralized in the responsible class. Would be sufficient for now.

When we'll introduce the mapping for e.g. planned production one can further refactor (Reuse Item Stock for Planned Production, optimize to either pull all or specific as needed).